### ISSUE TITLE
[Bug] FLASH_ATTN mishandles cross-attention key-padding masks when query_len != key_len

### ISSUE THREAD

**Issue by cr-gao** [vllm-project/vllm-omni#5865]
## Summary

On the `FLASH_ATTN` diffusion attention backend, a cross-attention call with a
2D key-padding mask and `query_len != key_len` zeroes out most query positions
in the output. The same inputs are handled correctly by `TORCH_SDPA` and
`CUDNN_ATTN`. This lives in shared attention infrastructure, so it affects any
diffusion model that runs prompt cross-attention with a padded key mask on
FLASH_ATTN, not just one model.

## Affected component

The FLASH varlen-masked path (`FlashAttentionImpl._forward_varlen_masked` ->
HF-derived `_upad_input` in `backends/utils/fa.py`) is written for
self-attention, where `query_len == key_len`. In cross-attention the image
query (thousands of tokens) is much longer than the text key (a few hundred).
The `_upad_input` else-branch slices the mask with `attention_mask[:, -query_length:]`
(assuming left padding on a self-attention sequence) and then unpads the
*image query* with the *text* key mask. Only the query rows whose index happens
to fall on a True mask position survive; `_pad_input` scatters everything else
back as zeros, so the prompt conditioning is effectively dropped.

With `batch_size >= 2` it is worse: the kept indices are computed on the
flattened `[batch, key_len]` mask but applied to the flattened
`[batch, query_len]` query, so the surviving rows are not even the leading
rows of each batch element.

## Impact

Discovered on SANA-Video 2B (native pipelines) on A800 (Ampere), where the
platform default backend is FLASH_ATTN. A single transformer forward on real
480p weights (tp=1), compared against a frozen golden output tensor:

| backend    | cosine  | rel_l2 | max_abs | vs thresholds (0.999 / 0.017 / 0.2) |
|------------|---------|--------|---------|-------------------------------------|
| FLASH_ATTN | 0.9157  | 0.412  | 8.64    | FAIL                                |
| TORCH_SDPA | 0.99993 | 0.0122 | 0.086   | PASS                                |
| CUDNN_ATTN | 0.99991 | 0.0141 | 0.125   | PASS                                |

Over 50 CFG denoising steps this compounds into unusable video (480p T2V lands
at ~0.29 SSIM on FLASH_ATTN vs ~0.94 on TORCH_SDPA / CUDNN_ATTN).

## Steps to reproduce (no model weights required)

Run a single cross-attention forward with `query_len != key_len` and a padded
key mask through FLASH_ATTN, and compare against `scaled_dot_product_attention`:

```python
import torch
import torch.nn.functional as F

torch.manual_seed(0)
device = "cuda"
b, heads, dim = 1, 8, 64
q_len, k_len = 4096, 300          # image query vs text key: q_len != k_len
q = torch.randn(b, heads, q_len, dim, device=device, dtype=torch.bfloat16)
k = torch.randn(b, heads, k_len, dim, device=device, dtype=torch.bfloat16)
v = torch.randn(b, heads, k_len, dim, device=device, dtype=torch.bfloat16)

# key-padding: only the first 120 text tokens are valid.
keep = torch.zeros(b, k_len, dtype=torch.bool, device=device)
keep[:, :120] = True
attn_mask = keep[:, None, None, :]            # broadcast over heads and query

ref = F.scaled_dot_product_attention(q, k, v, attn_mask=attn_mask)

# Route the identical q/k/v/keep-mask through the vLLM-Omni FLASH_ATTN
# cross-attention path (OmniAttention, role='cross', backend forced to
# FLASH_ATTN) and diff against `ref`.
```

## Expected vs actual

- **Expected:** the FLASH_ATTN output matches the SDPA/CUDNN output; every query
  row carries a real value.
- **Actual:** `max_abs` diff ~2.08; in the FLASH_ATTN output most query rows
  are ~0, because the key mask is (incorrectly) used to unpad the query.

## Environment

- GPU: NVIDIA A800-SXM4-80GB (Ampere), also reproduced on RTX 4070 (Ada)
- Platform default diffusion attention backend on these GPUs: FLASH_ATTN
- Model where it surfaced: SANA-Video 2B native T2V / I2V pipelines
- Correct backends: TORCH_SDPA, CUDNN_ATTN

## Suggested fix

In this scenario the query has no padding at all — the mask describes keys
only. So when `query_len != key_len`, keep the query dense and unpad only
K/V with the key mask, instead of reusing the key mask for the query.
Alternatively, route cross-attention with a key-only padding mask onto the
dense masked kernel that TORCH_SDPA / CUDNN_ATTN use.

## Workaround

Run the affected model on `TORCH_SDPA` or `CUDNN_ATTN` via
`--diffusion-attention-backend`. On Blackwell the platform default already
selects `CUDNN_ATTN`, so the issue does not surface there; it appears on GPUs
whose default resolves to `FLASH_ATTN` (e.g. Ampere).

**Comment by cr-gao** [comment]
Empirical confirmation of the `batch_size >= 2` claim (pre-fix code, b=2, q_len=256, k_len=40, 17 valid keys per row):

- batch element 1's output is **entirely zero** (all 256 query rows);
- batch element 0 keeps only 34 nonzero rows: 0–16 and 40–56 (17 = valid keys, 40 = k_len — the flattened key-mask indices applied verbatim to the query);
- rows 40–56 of batch 0 match SDPA computed with **batch 1's** K/V (max diff 0.0078, bf16 noise) and not batch 0's own K/V (diff 2.66).

So with batching this is not just dropped conditioning — it is cross-sample contamination: query rows attend to another sample's keys.

Also reproduced the model-level impact on A800 with real SANA-Video 2B 480p weights (single transformer forward vs frozen golden): FLASH_ATTN gives cosine 0.9157 / rel_l2 0.412 / max_abs 8.64, matching the table above bit-for-bit.
